misc-utils/fadvise.c

```diff
@@ -0,0 +1,181 @@
+	fprintf(out, _(" %s [options] file\n"), program_invocation_short_name);
+	fprintf(out, _(" %s [options] --fd|-d file-descriptor\n"), program_invocation_short_name);
+
+	fputs(USAGE_OPTIONS, out);
+	fputs(_(" -a, --advice <advice> applying advice to the file (default: \"dontneed\")\n"), out);
+	fputs(_(" -l, --length <num>    length for range operations, in bytes\n"), out);
+	fputs(_(" -o, --offset <num>    offset for range operations, in bytes\n"), out);
+
+	fputs(USAGE_SEPARATOR, out);
+	printf(USAGE_HELP_OPTIONS(23));
+
+	fputs(_("\nAvailable values for advice:\n"), out);
+	for (i = 0; i < ARRAY_SIZE(advices); i++) {
+		fprintf(out, "  %s\n",
+			advices[i].name);
+	}
+
+	printf(USAGE_MAN_TAIL("fadvise(1)"));
+
+	exit(EXIT_SUCCESS);
+}
+
+struct fadvise_args {
+	int fd;
+	off_t offset;
+	off_t len;
+	int advice;
+};
+
+static int fadvise_args(struct fadvise_args *args)
+{
+	return posix_fadvise(args->fd,
+			     args->offset, args->len,
+			     args->advice);
+}
+
+/* Copied from fallocate.c. */
+static loff_t cvtnum(char *s)
+{
+	uintmax_t x;
+
+	if (strtosize(s, &x))
+		return -1LL;
+
+	return x;
+}
+
+int main(int argc, char ** argv)
+{
+	int c;
+	int rc;
+	bool do_close = false;
+
+	struct fadvise_args args = {
+		.fd     = -1,
+		.offset = 0,
+		.len    = 0,
+		.advice = POSIX_FADV_DONTNEED,
+	};
+
+	static const struct option longopts[] = {
+		{ "advice",     required_argument, NULL, 'a' },
+		{ "fd",         required_argument, NULL, 'd' },
+		{ "length",     required_argument, NULL, 'l' },
+		{ "offset",     required_argument, NULL, 'o' },
+		{ "version",    no_argument,       NULL, 'V' },
+		{ "help",	no_argument,       NULL, 'h' },
+		{ NULL, 0, NULL, 0 },
+	};
+
+	setlocale(LC_ALL, "");
+	bindtextdomain(PACKAGE, LOCALEDIR);
+	textdomain(PACKAGE);
+
+	while ((c = getopt_long (argc, argv, "a:d:hl:o:", longopts, NULL)) != -1) {
+		switch (c) {
+		case 'a':
+			break;
+		case 'd':
+			args.fd = (int)cvtnum(optarg);
+			if (args.fd < 0)
+				errx(EXIT_FAILURE, _("invalid file descriptor value specified"));
+			break;
+		case 'l':
+			args.len = cvtnum(optarg);
+			if (args.len < 0)
+				errx(EXIT_FAILURE, _("invalid length value specified"));
+			break;
+		case 'o':
+			args.offset = cvtnum(optarg);
+			if (args.offset < 0)
+				errx(EXIT_FAILURE, _("invalid offset value specified"));
+			break;
+		case 'V':
+			print_version(EXIT_SUCCESS);
+		case 'h':
+			usage();
+		default:
+			errtryhelp(EXIT_FAILURE);
+		}
+	}
+
+	if (optind == argc && args.fd == -1) {
+		warnx(_("no file specified"));
+		errtryhelp(EXIT_FAILURE);
+	}
+
+	if (argc - optind > 0 && args.fd != -1) {
+		warnx(_("specify either file descriptor or file name"));
+		errtryhelp(EXIT_FAILURE);
+	}
+
+	if (argc - optind > 1) {
+		warnx(_("specify one file descriptor or file name"));
+		errtryhelp(EXIT_FAILURE);
+	}
+
+	if (args.fd == -1) {
+		args.fd = open(argv[optind], O_RDONLY);
+		if (args.fd < 0)
+			err(EXIT_FAILURE, _("failed to open: %s"), argv[optind]);
```

Comment:
We usually use "cannot open %s"   (see po/util-linux.pot for more details).

Reply:
Updated.